Kernel/Syscalls/jail.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Userspace.h>
+#include <Kernel/Jail.h>
+#include <Kernel/JailManagement.h>
+#include <Kernel/Process.h>
+#include <Kernel/StdLib.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+namespace Kernel {
+
+constexpr size_t jail_name_max_size = 50;
+
+ErrorOr<FlatPtr> Process::sys$jail_create(Userspace<Syscall::SC_jail_create_params*> user_params)
+{
+    VERIFY_NO_PROCESS_BIG_LOCK(this);
+    TRY(require_promise(Pledge::jail));
+
+    auto params = TRY(copy_typed_from_user(user_params));
+    auto jail_name = TRY(get_syscall_path_argument(params.name));
+    if (jail_name->length() > jail_name_max_size)
+        return ENAMETOOLONG;
+
+    auto jail = TRY(JailManagement::the().create_jail(move(jail_name)));
+    params.index = jail->index().value();
+
+    TRY(copy_to_user(user_params, &params));
+    return 0;
+}
+
+ErrorOr<FlatPtr> Process::sys$jail_attach(Userspace<Syscall::SC_jail_attach_params const*> user_params)
```

Comment:
This interface requires a process to add itself to a jail initially. So, this is something that must be done just after fork(), before exec? What about in the posix_spawn case? 

Should we add a non-portable posix_spawnattr item to add the new process to a jail? 

What about adding a new process to a jail externally? Should it be possible for a process to send another process to jail?

Reply:
In this case, I'm thinking about a server like `dockerd` that will manage a set of applications each in their own jail that might like to live itself in a non-jailed state but posix_spawn off some processes with a jail, and occasionally run commands in the context of that jail.

Reply:
Look at the `jail-attach` utility - it doesn't do any sort of `fork()` there and still enters into the a jail. You could also enter a jail and then do fork+exec and you still will be in a jail, and it also works the same if you join a jail after `fork()` but before `exec()`.